File: gwc.native/Sprite.cpp
/*
 * :.:.:.:.:.:.:.:.
 * GWC
 * Graphical Window
 * for Console Apps
 * :.:.:.:.:.:.:.:.
 *
 * GWC Native
 *
 * A Graphics Library
 *
 * https://github.com/reallukee/gwc
 *
 * Nome file : Sprite.cpp
 *
 * Titolo    : SPRITE
 * Sommario  : Sprite
 *
 * Autore    : Luca Pollicino
 *             (https://github.com/reallukee)
 * Versione  : v0.6.3
 *             NOTA BENE: Campo INDICATIVO!
 * Licenza   : MIT
 */

#include "gwc.clr.hpp"

#include "Sprite.hpp"
#include "SpriteHelper.hpp"
#include "macros.hpp"

#include "ImageHelper.hpp"
#include "IconHelper.hpp"

#include "Canvas.hpp"
#include "CanvasHelper.hpp"

#ifdef __cplusplus

namespace gwc
{
    Sprite::Sprite(int width, int height)
    {
        SpriteHost* host = new SpriteHost(width, height);

        sprite = static_cast<CLRSpriteHost>(host);
    }

    Sprite::Sprite()
    {
        SpriteHost* host = new SpriteHost();

        sprite = static_cast<CLRSpriteHost>(host);
    }

    Sprite::~Sprite()
    {
        SpriteHost* host = static_cast<SpriteHost*>(sprite);

        delete host;
    }



    bool Sprite::isInitialized()
    {
        SpriteHost* host = static_cast<SpriteHost*>(sprite);

        if (host == nullptr)
        {
            return false;
        }

        return !host->isNull();
    }



    gColor Sprite::getBorderColor()
    {
        CCI_GET_COLOR_CPP(SpriteHost, sprite, BorderColor);
    }

    void Sprite::setBorderColor(gColor color)
    {
        CCI_SET_COLOR_CPP(SpriteHost, sprite, BorderColor, color);
    }

    gColor Sprite::getFillColor()
    {
        CCI_GET_COLOR_CPP(SpriteHost, sprite, FillColor);
    }

    void Sprite::setFillColor(gColor color)
    {
        CCI_SET_COLOR_CPP(SpriteHost, sprite, FillColor, color);
    }

    gColor Sprite::getBackColor()
    {
        CCI_GET_COLOR_CPP(SpriteHost, sprite, BackColor);
    }

    void Sprite::setBackColor(gColor color)
    {
        CCI_SET_COLOR_CPP(SpriteHost, sprite, BackColor, color);
    }



    gRectangle Sprite::bounds()
    {
        CCI_GET_RECTANGLE_CPP(SpriteHost, sprite, Bounds);
    }

    gSize Sprite::size()
    {
        CCI_GET_SIZE_CPP(SpriteHost, sprite, Size);
    }

    gPoint Sprite::location()
    {
        CCI_GET_POINT_CPP(SpriteHost, sprite, Location);
    }



    int Sprite::getWidth()
    {
        CCI_INT_CPP(SpriteHost, sprite, Width);
    }

    int Sprite::getHeight()
    {
        CCI_INT_CPP(SpriteHost, sprite, Height);
    }



    bool Sprite::clear()
    {
        CCI_BOOL_CPP(SpriteHost, sprite, Clear());
    }

    bool Sprite::clearArea(int x, int y, int width, int height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, ClearArea(x, y, width, height));
    }

    bool Sprite::clearF()
    {
        CCI_BOOL_CPP(SpriteHost, sprite, ClearF());
    }

    bool Sprite::clearAreaF(float x, float y, float width, float height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, ClearAreaF(x, y, width, height));
    }



    bool Sprite::drawCanvas(int x, int y, const Canvas* canvas)
    {
        CanvasHost* canvasHost = CanvasHelper::get(canvas);

        GWC::Canvas^ managedCanvas = canvasHost->invoke();

        CCI_BOOL_CPP(SpriteHost, sprite, DrawCanvas(x, y, managedCanvas));
    }

    bool Sprite::drawCanvasF(float x, float y, const Canvas* canvas)
    {
        CanvasHost* canvasHost = CanvasHelper::get(canvas);

        GWC::Canvas^ managedCanvas = canvasHost->invoke();

        CCI_BOOL_CPP(SpriteHost, sprite, DrawCanvasF(x, y, managedCanvas));
    }



    bool Sprite::drawSprite(int x, int y, const Sprite* sprite)
    {
        SpriteHost* spriteHost = SpriteHelper::get(sprite);

        GWC::Sprite^ managedSprite = spriteHost->invoke();

        CCI_BOOL_CPP(SpriteHost, this->sprite, DrawSprite(x, y, managedSprite));
    }

    bool Sprite::drawSpriteF(float x, float y, const Sprite* sprite)
    {
        SpriteHost* spriteHost = SpriteHelper::get(sprite);

        GWC::Sprite^ managedSprite = spriteHost->invoke();

        CCI_BOOL_CPP(SpriteHost, this->sprite, DrawSpriteF(x, y, managedSprite));
    }



    bool Sprite::drawBorderSquare(int x, int y, int side)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderSquare(x, y, side));
    }

    bool Sprite::drawBorderSquareF(float x, float y, float side)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderSquareF(x, y, side));
    }

    bool Sprite::drawFillSquare(int x, int y, int side)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillSquare(x, y, side));
    }

    bool Sprite::drawFillSquareF(float x, float y, float side)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillSquareF(x, y, side));
    }



    bool Sprite::drawBorderRectangle(int x, int y, int width, int height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderRectangle(x, y, width, height));
    }

    bool Sprite::drawBorderRectangleF(float x, float y, float width, float height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderRectangleF(x, y, width, height));
    }

    bool Sprite::drawFillRectangle(int x, int y, int width, int height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillRectangle(x, y, width, height));
    }

    bool Sprite::drawFillRectangleF(float x, float y, float width, float height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillRectangleF(x, y, width, height));
    }



    bool Sprite::drawBorderCircle(int x, int y, int radius)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderCircle(x, y, radius));
    }

    bool Sprite::drawBorderCircleF(float x, float y, float radius)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderCircleF(x, y, radius));
    }

    bool Sprite::drawFillCircle(int x, int y, int radius)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillCircle(x, y, radius));
    }

    bool Sprite::drawFillCircleF(float x, float y, float radius)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillCircleF(x, y, radius));
    }



    bool Sprite::drawBorderEllipse(int x, int y, int width, int height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderEllipse(x, y, width, height));
    }

    bool Sprite::drawBorderEllipseF(float x, float y, float width, float height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawBorderEllipseF(x, y, width, height));
    }

    bool Sprite::drawFillEllipse(int x, int y, int width, int height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillEllipse(x, y, width, height));
    }

    bool Sprite::drawFillEllipseF(float x, float y, float width, float height)
    {
        CCI_BOOL_CPP(SpriteHost, sprite, DrawFillEllipseF(x, y, width, height));
    }



    static System::Drawing::Bitmap^ toManagedBitmap(Gdiplus::Bitmap* nativeBitmap)
    {
        if (!nativeBitmap)
        {
            return nullptr;
        }

        UINT width  = nativeBitmap->GetWidth();
        UINT height = nativeBitmap->GetHeight();

        Gdiplus::Rect nativeRect(
            0,
            0,
            width,
            height
        );

        Gdiplus::BitmapData nativeBitmapData;

        bool result = nativeBitmap->LockBits(
            &nativeRect,
            Gdiplus::ImageLockModeRead,
            PixelFormat32bppPARGB,
            &nativeBitmapData
        );

        if (result != Gdiplus::Status::Ok)
        {
            return nullptr;
        }

        System::Drawing::Bitmap^ managedBitmap = nullptr;

        try
        {
            managedBitmap = gcnew System::Drawing::Bitmap(
                width,
                height,
                nativeBitmapData.Stride,
                System::Drawing::Imaging::PixelFormat::Format32bppPArgb,
                (IntPtr)nativeBitmapData.Scan0
            );

            System::Drawing::Rectangle managedRectangle(
                0,
                0,
                width,
                height
            );

            System::Drawing::Bitmap^ copy = managedBitmap->Clone(
                managedRectangle,
                System::Drawing::Imaging::PixelFormat::Format32bppPArgb
            );

            delete managedBitmap;

            managedBitmap = copy;
        }
        finally
        {
            nativeBitmap->UnlockBits(&nativeBitmapData);
        }

        return managedBitmap;
    }

    static Bitmap* getNativeImage(const gImage& image)
    {
        if (!image.isLoaded())
        {
            return false;
        }

        Bitmap* nativeImage = ImageHelper::get(image);

        if (nativeImage == nullptr)
        {
            return false;
        }

        return nativeImage;
    }

    static HICON getNativeIcon(const gIcon& icon)
    {
        if (!icon.isLoaded())
        {
            return false;
        }

        HICON nativeIcon = IconHelper::get(icon);

        if (nativeIcon == nullptr)
        {
            return false;
        }

        return nativeIcon;
    }



    bool Sprite::drawImage(int x, int y, const gImage& image)
    {
        CC_BOOL_CPP(SpriteHost, sprite);

        Bitmap* nativeImage = getNativeImage(image);

        System::Drawing::Image^ managedImage = nullptr;

        bool result = false;

        try
        {
            managedImage = toManagedBitmap(nativeImage);

            result = _host->invoke()->DrawImage(x, y, managedImage);
        }
        finally
        {
            if (managedImage)
            {
                delete managedImage;
            }
        }

        return result;
    }

    bool Sprite::drawImageF(float x, float y, const gImage& image)
    {
        CC_BOOL_CPP(SpriteHost, sprite);

        Bitmap* nativeImage = getNativeImage(image);

        System::Drawing::Image^ managedImage = nullptr;

        bool result = false;

        try
        {
            managedImage = toManagedBitmap(nativeImage);

            result = _host->invoke()->DrawImageF(x, y, managedImage);
        }
        finally
        {
            if (managedImage)
            {
                delete managedImage;
            }
        }

        return result;
    }

    bool Sprite::drawIcon(int x, int y, const gIcon& icon)
    {
        CC_BOOL_CPP(SpriteHost, sprite);

        HICON nativeIcon = getNativeIcon(icon);

        System::Drawing::Icon^ managedIcon = nullptr;

        bool result = false;

        try
        {
            IntPtr iconHandle = IntPtr(nativeIcon);

            managedIcon = System::Drawing::Icon::FromHandle(iconHandle);

            result = _host->invoke()->DrawIcon(x, y, managedIcon);
        }
        finally
        {
            if (managedIcon)
            {
                delete managedIcon;
            }
        }

        return result;
    }

    bool Sprite::drawIconF(float x, float y, const gIcon& icon)
    {
        CC_BOOL_CPP(SpriteHost, sprite);

        HICON nativeIcon = getNativeIcon(icon);

        System::Drawing::Icon^ managedIcon = nullptr;

        bool result = false;

        try
        {
            IntPtr iconHandle = IntPtr(nativeIcon);

            managedIcon = System::Drawing::Icon::FromHandle(iconHandle);

            result = _host->invoke()->DrawIconF(x, y, managedIcon);
        }
        finally
        {
            if (managedIcon)
            {
                delete managedIcon;
            }
        }

        return result;
    }



    void Sprite::render()
    {
        CCI_VOID_CPP(SpriteHost, sprite, Render());
    }



    SpriteHost* SpriteHelper::get(const Sprite* sprite)
    {
        SpriteHost* spriteHost = static_cast<SpriteHost*>(sprite->sprite);

        return spriteHost;
    }
}

#endif // __cplusplus

File: gwc/src/Sprite.cs
/*
 * :.:.:.:.:.:.:.:.
 * GWC
 * Graphical Window
 * for Console Apps
 * :.:.:.:.:.:.:.:.
 *
 * A Graphics Library
 *
 * https://github.com/reallukee/gwc
 *
 * Nome file : Sprite.cs
 *
 * Titolo    : SPRITE
 * Sommario  : Contiene l'implementazione della
 *             classe Sprite.
 *
 * Autore    : Luca Pollicino
 *             (https://github.com/reallukee)
 * Versione  : v0.6.3
 *             NOTA BENE: Campo INDICATIVO!
 * Licenza   : MIT
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Resources;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading;

using System.Drawing;
using System.Drawing.Imaging;
using System.Windows.Forms;

using Reallukee.GWC.GPipeline;
using Reallukee.GWC.Internal;
using Reallukee.GWC.Interop;

namespace Reallukee.GWC
{
    public sealed class Sprite
        : IDisposable, IGRenderable, IGBorderColor, IGFillColor, IGBackColor
    {
        public const int MaxBufferLength = 10000;

        private bool disposed;

        public Sprite(int width, int height)
        {
            ThrowIfArgumentOutOfRange(
                nameof(width),
                width,
                IsLessOrEqualThen(0)
            );

            ThrowIfArgumentOutOfRange(
                nameof(height),
                height,
                IsLessOrEqualThen(0)
            );

            InitBitmap(width, height);
            InitBuffer(width, height);

            BorderColor = Color.Black;
            FillColor   = Color.Green;

            disposed = false;
        }

        public Sprite()
            : this(800, 600)
        {

        }



        public void Dispose()
        {
            Dispose(true);
        }

        private void Dispose(bool disposing)
        {
            if (disposed)
            {
                return;
            }

            if (disposing)
            {
                Bitmap?.Dispose();
                Marshal.FreeHGlobal(RawBitmap);

                Bitmap = null;
                RawBitmap = IntPtr.Zero;
            }

            disposed = true;
        }



        private delegate (bool Invalid, string Message) RangeCheck(int value);

        private RangeCheck IsLessOrEqualThen(int min)
        {
            return value =>
            {
                if (value <= min)
                {
                    return (
                        true,
                        $"Value cannot be less or equal than {min}."
                    );
                }

                return (false, null);
            };
        }

        private RangeCheck IsGreaterOrEqualThan(int max)
        {
            return value =>
            {
                if (value >= max)
                {
                    return (
                        true,
                        $"Value cannot be greater or equal than {max}."
                    );
                }

                return (false, null);
            };
        }

        private void ThrowIfArgumentOutOfRange(
            string name, int value, params RangeCheck[] rangeChecks
        )
        {
            foreach (var rangeCheck in rangeChecks)
            {
                (bool invalid, string message) = rangeCheck(value);

                if (invalid)
                {
                    throw new ArgumentOutOfRangeException(
                        name,
                        message
                    );
                }
            }
        }



        private object bitmapLock;

        private object bufferLock;

        private void InitBitmap(int width, int height)
        {
            try
            {
                RawBitmap = Marshal.AllocHGlobal(width * height * 4);

                Bitmap = new Bitmap(
                    width,
                    height,
                    width * 4,
                    PixelFormat.Format32bppPArgb,
                    RawBitmap
                );
            }
            catch (Exception ex)
            {
                MessageBox.Show(
                    $"Critical Error!\n Reason: {ex.Message}",
                    "GWC",
                    MessageBoxButtons.OK,
                    MessageBoxIcon.Error
                );
            }

            bitmapLock = new object();
        }

        private void InitBuffer(int width, int height)
        {
            Buffer = new ConcurrentQueue<IGRenderable>();

            bufferLock = new object();
        }



        internal int RawBitmapSize
        {
            get;
            set;
        }

        internal IntPtr RawBitmap
        {
            get;
            set;
        }

        internal Bitmap Bitmap
        {
            get;
            set;
        }

        internal ConcurrentQueue<IGRenderable> Buffer
        {
            get;
            set;
        }



        public Color BorderColor
        {
            get;
            set;
        }

        public Color FillColor
        {
            get;
            set;
        }

        public Color BackColor
        {
            get;
            set;
        }



        public Rectangle Bounds   => new Rectangle(0, 0, Width, Height);
        public Size      Size     => new Size     (Width, Height);
        public Point     Location => new Point    (0, 0);



        public int Width
        {
            get
            {
                lock (bitmapLock)
                {
                    return Bitmap.Width;
                }
            }
        }

        public int Height
        {
            get
            {
                lock (bitmapLock)
                {
                    return Bitmap.Height;
                }
            }
        }



        internal bool IsCached
        {
            get;
            private set;
        }

        internal bool DrawRenderable(IGRenderable renderable)
        {
            if (Buffer.Count > MaxBufferLength)
            {
                return false;
            }

            Buffer.Enqueue(renderable);

            IsCached = false;

            return true;
        }



        public bool Clear()
        {
            IGRenderable renderable = new GClear(BackColor);

            return DrawRenderable(renderable);
        }

        public bool ClearArea(int x, int y, int width, int height)
        {
            IGRenderable renderable = new GClear(x, y, width, height, BackColor);

            return DrawRenderable(renderable);
        }

        public bool ClearF()
        {
            IGRenderable renderable = new GClearF(BackColor);

            return DrawRenderable(renderable);
        }

        public bool ClearAreaF(float x, float y, float width, float height)
        {
            IGRenderable renderable = new GClearF(x, y, width, height, BackColor);

            return DrawRenderable(renderable);
        }



        public bool DrawCanvas(Canvas canvas)
        {
            IGRenderable renderable = new GCanvas(0, 0, canvas);

            return DrawRenderable(renderable);
        }

        public bool DrawCanvas(int x, int y, Canvas canvas)
        {
            IGRenderable renderable = new GCanvas(x, y, canvas);

            return DrawRenderable(renderable);
        }

        public bool DrawCanvasF(float x, float y, Canvas canvas)
        {
            IGRenderable renderable = new GCanvasF(x, y, canvas);

            return DrawRenderable(renderable);
        }



        public bool DrawSprite(Sprite sprite)
        {
            IGRenderable renderable = new GSprite(0, 0, sprite);

            return DrawRenderable(renderable);
        }

        public bool DrawSprite(int x, int y, Sprite sprite)
        {
            IGRenderable renderable = new GSprite(x, y, sprite);

            return DrawRenderable(renderable);
        }

        public bool DrawSpriteF(float x, float y, Sprite sprite)
        {
            IGRenderable renderable = new GSpriteF(x, y, sprite);

            return DrawRenderable(renderable);
        }



        public bool DrawBorderSquare(int x, int y, int side)
        {
            return DrawBorderRectangle(x, y, side, side);
        }

        public bool DrawBorderSquareF(float x, float y, float side)
        {
            return DrawBorderRectangleF(x, y, side, side);
        }

        public bool DrawFillSquare(int x, int y, int side)
        {
            return DrawFillRectangle(x, y, side, side);
        }

        public bool DrawFillSquareF(float x, float y, float side)
        {
            return DrawFillRectangleF(x, y, side, side);
        }



        public bool DrawBorderRectangle(int x, int y, int width, int height)
        {
            IGRenderable renderable = new GBorderRectangle(BorderColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawBorderRectangleF(float x, float y, float width, float height)
        {
            IGRenderable renderable = new GBorderRectangleF(BorderColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawFillRectangle(int x, int y, int width, int height)
        {
            IGRenderable renderable = new GFillRectangle(FillColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawFillRectangleF(float x, float y, float width, float height)
        {
            IGRenderable renderable = new GFillRectangleF(FillColor, x, y, width, height);

            return DrawRenderable(renderable);
        }



        public bool DrawBorderCircle(int x, int y, int radius)
        {
            return DrawBorderEllipse(x, y, radius * 2, radius * 2);
        }

        public bool DrawBorderCircleF(float x, float y, float radius)
        {
            return DrawBorderEllipseF(x, y, radius * 2, radius * 2);
        }

        public bool DrawFillCircle(int x, int y, int radius)
        {
            return DrawFillEllipse(x, y, radius * 2, radius * 2);
        }

        public bool DrawFillCircleF(float x, float y, float radius)
        {
            return DrawFillEllipseF(x, y, radius * 2, radius * 2);
        }



        public bool DrawBorderEllipse(int x, int y, int width, int height)
        {
            IGRenderable renderable = new GBorderEllipse(BorderColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawBorderEllipseF(float x, float y, float width, float height)
        {
            IGRenderable renderable = new GBorderEllipseF(BorderColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawFillEllipse(int x, int y, int width, int height)
        {
            IGRenderable renderable = new GFillEllipse(FillColor, x, y, width, height);

            return DrawRenderable(renderable);
        }

        public bool DrawFillEllipseF(float x, float y, float width, float height)
        {
            IGRenderable renderable = new GFillEllipseF(FillColor, x, y, width, height);

            return DrawRenderable(renderable);
        }



        public bool DrawImage(int x, int y, Image image)
        {
            IGRenderable renderable = new GImage(x, y, image);

            return DrawRenderable(renderable);
        }

        public bool DrawImageF(float x, float y, Image image)
        {
            IGRenderable renderable = new GImageF(x, y, image);

            return DrawRenderable(renderable);
        }

        public bool DrawIcon(int x, int y, Icon icon)
        {
            IGRenderable renderable = new GIcon(x, y, icon);

            return DrawRenderable(renderable);
        }

        public bool DrawIconF(float x, float y, Icon icon)
        {
            IGRenderable renderable = new GIconF(x, y, icon);

            return DrawRenderable(renderable);
        }



        public void Render(Graphics g)
        {
            if (IsCached)
            {
                lock (bitmapLock)
                {
                    g.DrawImage(Bitmap, 0, 0);
                }
            }
            else
            {
                while (Buffer.TryDequeue(out IGRenderable renderable))
                {
                    renderable.Render(g);

                    if (renderable is IDisposable disposable)
                    {
                        disposable.Dispose();
                    }
                }

                IsCached = true;
            }
        }

        public void Render()
        {
            lock (bitmapLock)
            {
                using (Graphics g = Graphics.FromImage(Bitmap))
                {
                    Render(g);
                }

                IsCached = true;
            }
        }
    }
}
